lib/backend/consul/consul.go

```diff
@@ -0,0 +1,613 @@
+	}
+
+	client, err := consulapi.NewClient(&cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &ClientWrapper{Client: client}, nil
+}
+
+func getReadyFlagPair() *model.KVPair {
+	return &model.KVPair{
+		Key:   model.ReadyFlagKey{},
+		Value: true,
+	}
+}
+
+// EnsureInitialized makes sure that the etcd data is initialized for use by
+// Calico.
+func (c *ClientWrapper) EnsureInitialized() error {
+	// Make sure the Ready flag is initialized in the datastore
+	if _, err := c.Create(getReadyFlagPair()); err == nil {
+		log.Info("Ready flag is now set")
+	} else {
+		if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
+			log.WithError(err).Warn("Failed to set ready flag")
+			return err
+		}
+		log.Info("Ready flag is already set")
+	}
+
+	return nil
+}
+
+func (c *ClientWrapper) Syncer(callbacks api.SyncerCallbacks) api.Syncer {
+	return newSyncer(c.Client, callbacks)
+}
+
+// Create an entry in the datastore.  This errors if the entry already exists.
+func (c *ClientWrapper) Create(d *model.KVPair) (*model.KVPair, error) {
+	return c.set(d, &setOptions{Kind: create})
+}
+
+// Update an existing entry in the datastore.  This errors if the entry does
+// not exist.
+func (c *ClientWrapper) Update(d *model.KVPair) (*model.KVPair, error) {
+	// If the request includes a revision, set it as the etcd previous index.
+	options := setOptions{Kind: update}
+	if d.Revision != nil {
+		options.Index = d.Revision.(uint64)
+		log.Debugf("Performing CAS against consulapi index: %v\n", options.Index)
+	}
+
+	return c.set(d, &options)
+}
+
+// Set an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *ClientWrapper) Apply(d *model.KVPair) (*model.KVPair, error) {
+	return c.set(d, &setOptions{Kind: replace})
+}
+
+// Delete an entry in the datastore.  This errors if the entry does not exists.
+func (c *ClientWrapper) Delete(d *model.KVPair) error {
+	path, err := keyToDefaultDeletePath(d.Key)
+	if err != nil {
+		return err
+	}
+	deleteOp := &consulapi.KVTxnOp{Key: path, Verb: consulapi.KVDeleteTree}
+	if d.Revision != nil {
+		deleteOp.Index = d.Revision.(uint64)
+	}
+	log.Debugf("Delete Key: %s", path)
+
+	kv := c.Client.KV()
+
+	readyFlagPair := getReadyFlagPair()
+	readyPath, err := keyToDefaultPath(readyFlagPair.Key)
+	if err != nil {
+		return err
+	}
+
+	serializedValue, err := model.SerializeValue(readyFlagPair)
+	if err != nil {
+		return err
+	}
+	updateReadyOp := &consulapi.KVTxnOp{Key: readyPath, Value: serializedValue, Verb: consulapi.KVSet}
+	ops := consulapi.KVTxnOps{deleteOp, updateReadyOp}
+	ok, response, _, err := kv.Txn(ops, nil)
+	if err != nil {
+		return convertConsulError(err, d.Key)
+	}
+
+	if !ok || len(response.Errors) > 0 {
+		return convertArrayOfErrorsToError(convertTxnErrors(d.Key, ops, response.Errors))
+	}
+
+	// If there are parents to be deleted, delete these as well provided there
+	// are no more children.
+	// TODO: investigate if we need to do it in consul
```

Comment:
As far as I can understand we need this in situation when there were keys `/a/b/c` and `/a/b/d`. So, children (c and d) got deleted and we want to clean up `/a/b'. For consul this is not a case, because if we didn't create `/a/b` explicitly it will return 404 on `/a/b`.
```
/ # consul kv put calico/v1/a/b/c 1
Success! Data written to: calico/v1/a/b/c
/ # consul kv get calico/v1/a/b
Error! No key exists at: calico/v1/a/b
```

But if we do create parents explicitly and logic of calico requires them to be cleaned up we will need to cleanup.